In [1]:
import pandas as pd

In [2]:
df = pd.read_csv("survey_results_public.csv", low_memory=False)
schema = pd.read_csv("survey_results_schema.csv")

df.head()

,ResponseId,MainBranch,Age,EdLevel,Employment,EmploymentAddl,WorkExp,LearnCodeChoose,LearnCode,LearnCodeAI,...,AIAgentOrchestration,AIAgentOrchWrite,AIAgentObserveSecure,AIAgentObsWrite,AIAgentExternal,AIAgentExtWrite,AIHuman,AIOpen,ConvertedCompYearly,JobSat
0,1,I am a developer by profession,25-34 years old,"Master’s degree (M.A., M.S., M.Eng., MBA, etc.)",Employed,"Caring for dependents (children, elderly, etc.)",8.0,"Yes, I am not new to coding but am learning ne...",Online Courses or Certification (includes all ...,"Yes, I learned how to use AI-enabled tools for...",...,Vertex AI,NaN,NaN,NaN,ChatGPT,NaN,When I don’t trust AI’s answers,"Troubleshooting, profiling, debugging",61256.0,10.0
1,2,I am a developer by profession,25-34 years old,"Associate degree (A.A., A.S., etc.)",Employed,NaN,2.0,"Yes, I am not new to coding but am learning ne...",Online Courses or Certification (includes all ...,"Yes, I learned how to use AI-enabled tools for...",...,NaN,NaN,NaN,NaN,NaN,NaN,When I don’t trust AI’s answers;When I want to...,All skills. AI is a flop.,104413.0,9.0
2,3,I am a developer by profession,35-44 years old,"Bachelor’s degree (B.A., B.S., B.Eng., etc.)","Independent contractor, freelancer, or self-em...",None of the above,10.0,"Yes, I am not new to coding but am learning ne...",Online Courses or Certification (includes all ...,"Yes, I learned how to use AI-enabled tools for...",...,NaN,NaN,NaN,NaN,ChatGPT;Claude Code;GitHub Copilot;Google Gemini,NaN,When I don’t trust AI’s answers;When I want to...,"Understand how things actually work, problem s...",53061.0,8.0
3,4,I am a developer by profession,35-44 years old,"Bachelor’s degree (B.A., B.S., B.Eng., etc.)",Employed,None of the above,4.0,"Yes, I am not new to coding but am learning ne...","Other online resources (e.g. standard search, ...","Yes, I learned how to use AI-enabled tools for...",...,NaN,NaN,NaN,NaN,ChatGPT;Claude Code,NaN,When I don’t trust AI’s answers;When I want to...,NaN,36197.0,6.0
4,5,I am a developer by profession,35-44 years old,"Master’s degree (M.A., M.S., M.Eng., MBA, etc.)","Independent contractor, freelancer, or self-em...","Caring for dependents (children, elderly, etc.)",21.0,"No, I am not new to coding and did not learn n...",NaN,"Yes, I learned how to use AI-enabled tools for...",...,NaN,NaN,NaN,NaN,NaN,NaN,When I don’t trust AI’s answers,"critical thinking, the skill to define the tas...",60000.0,7.0


In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 49191 entries, 0 to 49190
Columns: 172 entries, ResponseId to JobSat
dtypes: float64(52), int64(1), object(119)
memory usage: 64.6+ MB


Завдання 1. Підрахунок загальної кількості респондентів

In [4]:
total_respondents = df['ResponseId'].nunique()

print("Загальна кількість респондентів:", total_respondents)

Загальна кількість респондентів: 49191


Завдання 2. Аналіз повноти відповідей респондентів

In [5]:
schema_questions = set(schema['qname'])
df_columns = set(df.columns)

In [6]:
# Знаходимо спільні колонки
common_questions = schema_questions & df_columns

# Додаємо ResponseId вручну
common_questions.add('ResponseId')

# В список
common_questions = list(common_questions)

# Фільтруємо датасет
df_filtered = df[common_questions]

# Видаляємо рядки з пропусками
complete_answers = df_filtered.dropna()

# Рахуємо унікальних респондентів
complete_count = complete_answers['ResponseId'].nunique()

print("Кількість респондентів без пропусків:", complete_count)

Кількість респондентів без пропусків: 0


Завдання 3. Статистичний аналіз досвіду респондентів

In [7]:
work_exp = df['WorkExp'].dropna()

mean_exp = work_exp.mean()
median_exp = work_exp.median()
mode_exp = work_exp.mode()[0]

stats = pd.Series({
    "Mean": mean_exp,
    "Median": median_exp,
    "Mode": mode_exp})
stats

Mean      13.367403
Median    10.000000
Mode      10.000000
dtype: float64

Середній досвід респондентів становить приблизно 13 років, медіанний та модальний — 10 років, що свідчить про концентрацію більшості відповідей навколо цього значення.

Завдання 4. Аналіз віддаленої роботи

In [8]:
df['RemoteWork'].value_counts(dropna=False)

RemoteWork
NaN                                                                             15411
Remote                                                                          10931
Hybrid (some remote, leans heavy to in-person)                                   6732
In-person                                                                        6042
Hybrid (some in-person, leans heavy to flexibility)                              5831
Your choice (very flexible, you can come in when you want or just as needed)     4244
Name: count, dtype: int64

In [9]:
remote_count = df[df['RemoteWork'] == 'Remote']['ResponseId'].nunique()

print("Кількість респондентів, які працюють віддалено:", remote_count)

Кількість респондентів, які працюють віддалено: 10931


Завдання 5. Визначення популярності Python

In [10]:
pd.set_option('display.max_colwidth', None)
pd.set_option('display.max_columns', None)
pd.set_option('display.width', None)

In [11]:
schema[schema['qname'].str.contains('Language', case=False, na=False)][['qname', 'question']]

,qname,question
75,LanguageChoice,"Did you work with programming, scripting and markup languages in the last year or do you plan to in the year to come? Here, we are referring to anything you use that is conventionally referred to as code."
76,Language,"Which <b>programming, scripting, and markup languages</b> have you done extensive development work in over the past year, and which do you want to work in over the next year? (If you both worked with the language and want to continue to do so, please check both boxes in that row.)"
77,LanguagesHaveEntry,"Was the programming language you use not listed above? List below, separated by a comma and a space for each, the programming language(s) <strong>you used in the past year</strong>."
78,LanguagesWantEntry,"Was the programming language you want to use not listed above? List below, separated by a comma and a space for each, the programming language(s) <strong>you want to use next year</strong>."


In [12]:
df['LanguageHaveWorkedWith'].head(10)

0                                                                         Bash/Shell (all shells);Dart;SQL
1                                                                                                     Java
2                                                                      Dart;HTML/CSS;JavaScript;TypeScript
3                                                                                          Java;Kotlin;SQL
4    C;C#;C++;Delphi;HTML/CSS;Java;JavaScript;Lua;PowerShell;Python;SQL;TypeScript;VBA;Visual Basic (.Net)
5                                                                                               Java;Scala
6                                                                                    JavaScript;TypeScript
7                                        Bash/Shell (all shells);HTML/CSS;JavaScript;Python;SQL;TypeScript
8                                                                                        Java;Python;Scala
9                                    

In [13]:
[col for col in df.columns if 'lang' in col.lower()]

['LanguageChoice',
 'LanguageHaveWorkedWith',
 'LanguageWantToWorkWith',
 'LanguageAdmired',
 'LanguagesHaveEntry',
 'LanguagesWantEntry']

In [14]:
df['UsesPython'] = df['LanguageHaveWorkedWith'].str.contains('Python', case=False, na=False)

python_count = df[df['UsesPython']]['ResponseId'].nunique()

total_count = df['ResponseId'].nunique()

python_percent = (python_count / total_count) * 100

print(f"Відсоток респондентів, які програмують на Python: {python_percent:.2f}%")

Відсоток респондентів, які програмують на Python: 37.54%


Завдання 6. Аналіз шляхів навчання програмуванню

In [15]:
df['LearnCode'].value_counts(dropna=False)

LearnCode
NaN                                                                                                                                                                                                                                                                                                                                         15635
Other online resources (e.g. standard search, forum, online community);Stack Overflow or Stack Exchange;Technical documentation (is generated for/by the tool or system)                                                                                                                                                                      632
Other online resources (e.g. standard search, forum, online community);Technical documentation (is generated for/by the tool or system)                                                                                                                                                                                   

In [16]:
online_learners = df['LearnCode'].str.contains('Online Courses', case=False, na=False)

online_count = df[online_learners]['ResponseId'].nunique()

print("Кількість респондентів, які навчались через онлайн-курси:", online_count)

Кількість респондентів, які навчались через онлайн-курси: 10973


Завдання 7. Географічний аналіз компенсації Python-розробників

In [17]:
df[['Country', 'ConvertedCompYearly']].head()

,Country,ConvertedCompYearly
0,Ukraine,61256.0
1,Netherlands,104413.0
2,Ukraine,53061.0
3,Ukraine,36197.0
4,Ukraine,60000.0


In [18]:
python_devs = df[df['UsesPython']]

# Забираємо рядки без зарплати і країни
python_devs = python_devs.dropna(subset=['ConvertedCompYearly', 'Country'])

# Групуємо по країнах і рахуємо статистику
salary_by_country = (python_devs.groupby('Country')['ConvertedCompYearly'].agg(['mean', 'median']).reset_index())

# Перейменуємо колонки
salary_by_country.columns = ['Country', 'AverageCompensation', 'MedianCompensation']

# Сортуємо
salary_by_country = salary_by_country.sort_values(by='AverageCompensation', ascending=False)

salary_by_country.head(10)

,Country,AverageCompensation,MedianCompensation
102,Oman,390135.000000,390135.0
3,Andorra,226103.500000,226103.5
149,Viet Nam,218837.166667,8254.0
145,United States of America,173298.590211,150000.0
132,Switzerland,156456.600000,142592.0
121,Singapore,147515.121951,91391.0
65,Israel,135828.365385,142594.0
98,Nigeria,128773.276596,1808.0
63,Ireland,120523.918919,116015.0
99,Nomadic,120131.571429,139218.0


In [19]:
count_by_country = python_devs.groupby('Country')['ResponseId'].nunique()

salary_by_country['Count'] = count_by_country.values

salary_by_country.head(10)

,Country,AverageCompensation,MedianCompensation,Count
102,Oman,390135.000000,390135.0,3
3,Andorra,226103.500000,226103.5,5
149,Viet Nam,218837.166667,8254.0,7
145,United States of America,173298.590211,150000.0,2
132,Switzerland,156456.600000,142592.0,1
121,Singapore,147515.121951,91391.0,77
65,Israel,135828.365385,142594.0,5
98,Nigeria,128773.276596,1808.0,307
63,Ireland,120523.918919,116015.0,139
99,Nomadic,120131.571429,139218.0,3


Аналіз компенсації Python-розробників у розрізі країн показав, що найвищі середні показники мають країни з дуже малою кількістю респондентів (наприклад, Оман, Андорра, Швейцарія). Це свідчить про те, що на середнє значення в таких випадках суттєво впливає високий дохід окремих респондентів.

У країнах з більшою вибіркою (наприклад, Нігерія, Ірландія, Сінгапур) спостерігається значна різниця між середнім значенням та медіаною, що свідчить про нерівномірний розподіл доходів. У таких випадках медіана є більш надійним показником.

Тому при аналізі компенсації Python-розробників доцільно враховувати як статистичні показники, так і кількість учасників опитування.

Завдання 8. Аналіз освіти найбільш оплачуваних спеціалістів

In [20]:
df[['EdLevel', 'ConvertedCompYearly']].head()

,EdLevel,ConvertedCompYearly
0,"Master’s degree (M.A., M.S., M.Eng., MBA, etc.)",61256.0
1,"Associate degree (A.A., A.S., etc.)",104413.0
2,"Bachelor’s degree (B.A., B.S., B.Eng., etc.)",53061.0
3,"Bachelor’s degree (B.A., B.S., B.Eng., etc.)",36197.0
4,"Master’s degree (M.A., M.S., M.Eng., MBA, etc.)",60000.0


In [21]:
top_paid = df.dropna(subset=['ConvertedCompYearly', 'EdLevel'])

#Сортуємо та беремо топ-5
top_5 = top_paid.sort_values(by='ConvertedCompYearly', ascending=False).head(5)

top_5[['ConvertedCompYearly', 'EdLevel']]

,ConvertedCompYearly,EdLevel
34267,50000000.0,"Associate degree (A.A., A.S., etc.)"
28700,33552715.0,"Master’s degree (M.A., M.S., M.Eng., MBA, etc.)"
43143,18387548.0,"Associate degree (A.A., A.S., etc.)"
35353,15430267.0,"Bachelor’s degree (B.A., B.S., B.Eng., etc.)"
45971,13921760.0,"Master’s degree (M.A., M.S., M.Eng., MBA, etc.)"


У топ-5 найбільш високооплачуваних респондентів представлені різні рівні освіти — від Associate до Master’s degree, що свідчить про те, що у сфері програмування формальна освіта не є єдиним фактором успіху.

Завдання 9. Аналіз популярності Python по віковим категоріям

In [23]:
df['Age'].value_counts().head(10)

Age
25-34 years old      16519
35-44 years old      13241
18-24 years old       9210
45-54 years old       6275
55-64 years old       2626
65 years or older      942
Prefer not to say      378
Name: count, dtype: int64

In [24]:
age_df = df.dropna(subset=['Age'])

python_by_age = (age_df.groupby('Age').agg(total=('ResponseId', 'nunique'), python_users=('UsesPython', 'sum')).reset_index())

python_by_age['PythonPercent'] = (python_by_age['python_users'] / python_by_age['total']) * 100

python_by_age.sort_values('Age')

,Age,total,python_users,PythonPercent
0,18-24 years old,9210,3684,40.000000
1,25-34 years old,16519,6102,36.939282
2,35-44 years old,13241,4862,36.719281
3,45-54 years old,6275,2424,38.629482
4,55-64 years old,2626,978,37.242955
5,65 years or older,942,298,31.634820
6,Prefer not to say,378,118,31.216931


Аналіз популярності Python за віковими категоріями показав, що найвищий відсоток користувачів Python спостерігається серед респондентів віком 18–24 роки (близько 40%). Це свідчить про високу популярність Python серед початківців та молодих спеціалістів.

У вікових групах від 25 до 54 років частка Python-розробників залишається стабільною (приблизно 36–38%), що вказує на довгострокову актуальність мови у професійному середовищі.

У групі 65+ років відсоток дещо знижується, що може бути пов’язано з меншою готовністю до освоєння нових технологій.

Загалом, Python є популярною мовою програмування у всіх вікових групах, особливо серед молодих розробників.